api_subscription_management.go

```diff
@@ -2,26 +2,148 @@ package deribit
 
 import (
 	"context"
+	"strings"
 
 	"github.com/KyberNetwork/deribit-api/models"
 )
 
-func (c *Client) PublicSubscribe(ctx context.Context, params *models.SubscribeParams) (result models.SubscribeResponse, err error) {
+func (c *Client) Subscribe(channels []string) error {
+	return c.subscribe(channels, true)
+}
+
+func (c *Client) subscribe(channels []string, isNewSubscription bool) error {
+	l := c.l.With("func", "subscribe")
+	var publicChannels []string
+	var privateChannels []string
+
+	c.mu.RLock()
+	currentMap := c.subscriptionsMap
+	c.mu.RUnlock()
+	for _, v := range channels {
+		if _, ok := currentMap[v]; ok {
+			continue
+		}
+		if strings.HasPrefix(v, "user.") {
+			privateChannels = append(privateChannels, v)
+		} else {
+			publicChannels = append(publicChannels, v)
+		}
+	}
+
+	if len(publicChannels) > 0 {
+		pubSubResp, err := c.publicSubscribe(context.Background(), &models.SubscribeParams{
+			Channels: publicChannels,
+		})
+		if err != nil {
+			l.Errorw("error subscribe public", "err", err)
+			return err
+		}
+		c.mu.Lock()
+		if isNewSubscription {
+			c.subscriptions = append(c.subscriptions, pubSubResp...)
+		}
+		for _, v := range pubSubResp {
+			c.subscriptionsMap[v] = struct{}{}
+		}
+		c.mu.Unlock()
+	}
+
+	if len(privateChannels) > 0 {
+		privateSubResp, err := c.privateSubscribe(context.Background(), &models.SubscribeParams{
+			Channels: privateChannels,
+		})
+		if err != nil {
+			l.Errorw("error subscribe private", "err", err)
+			return err
+		}
+		c.mu.Lock()
+		if isNewSubscription {
+			c.subscriptions = append(c.subscriptions, privateSubResp...)
+		}
+		for _, v := range privateSubResp {
+			c.subscriptionsMap[v] = struct{}{}
+		}
+		c.mu.Unlock()
+	}
+	return nil
+}
+
+func (c *Client) UnSubscribe(channels []string) error {
+	l := c.l.With("func", "UnSubscribe")
+	var publicChannels []string
+	var privateChannels []string
+
+	c.mu.RLock()
+	currentMap := c.subscriptionsMap
```

Comment:
Same here This can have some race conditions when call in multiple go routines. I think we need to read and update the subscription mapping in one critical section.